doc/pserver.md

```diff
@@ -0,0 +1,261 @@
+2. Before each step of training, the worker pulls the latest model from KVStore.
+3. After a round of forward/backward computation, the worker pushes gradients to the Gradient Queue waiting for processing.
+4. Optimizer gets a gradient from the Gradient Queue.
+5. Then, Optimizer looks up the corresponding parameter from KVStore.
+6. Optimizer applies gradients to parameters, and updates parameter back to KVStore.
+
+
+### Tensor Data Structure
+
+To support data communication between pods, we introduce a `Tensor` proto message:
+
+```proto
+message Tensor {
+    enum DataType {
+        BOOL = 0;
+        INT16 = 1;
+        INT32 = 2;
+        INT64 = 3;
+        FP16 = 4;
+        FP32 = 5;
+        FP64 = 6;
+    }
+    string name = 1;
+    DataType data_type = 2;
+    repeated int64 dims = 3;
+    bytes content = 4;
+    repeated int64 indices = 5;
+}
+```
+
+Correspondingly, we have a `Tensor` Python class.
+
+
+```python
+class Tensor(object):
+    def __init__(self, name=None, value=None, indices=None):
+        self.name = name
+        self.value = value
+        self.indices = indices
+```
+
+There are also some helper functions:
+
+```python
+def serialize_to_pb(tensor, pb):
+    pass
+
+def deserialize_from_pb(pb, tensor):
+    pass
+
+def convert_to_tf_variable(tensor):
+    pass
+
+def convert_to_tf_tensor(tensor):
+    pass
+```
+
+
+
+### KVStore
+
+For a common model variable, we use save it as a `tf.Variable` in Parameter DB. For the embedding table, we introduce a customized data structure.
+
+
+```python
+class EmbeddingTable(object):
+    def __init__(self, name):
+        self.name = name
+        self.vectors = {}
+        
+    def get(self, indices):
+        ...
+        return tensor, unknown_indices
+        
+    def set(self, indices, value):
+        pass        
+```
+
+The name of embedding table is the embedding layer name. EmbeddingTable uses a dictionary `vectors` to store `<id, embedding_vector>` pairs.
+
+The KVStore could be like following:
+
+```python
+class KVStore(object):
+    def __init__(self):
+        self.param_db = {}
+        self.embedding_table_db = {}
+        
+    def get_param(self, name):
+        pass
+        
+    def get_embedding_vector(self, name, indices):
+        pass
+        
+    def set_param(self, name, value):
+        pass
+        
+    def set_embedding_vector(self, name, indices, value):
+        pass
+```
+
+### RPC Service
+
+KVStore and GradientQueue provide RPC service for workers.
+
+Since pserver will store a subset of the full model. And a worker will push/pull a submodel from the pserver. The model message is defined as following:
+
+```proto
+message Model {
+    int64 version = 1;
+    repeated Tensor tensors = 2;
+}
+```
+
+Model could also be used as gradients collection.
+
+So the RPC service will be defined as following:
+
+```proto
+message EmbeddingResponse{
+    Tensor value = 1;
+    repeated int64 unknown_indices = 2;
```

Comment:
No need to include unknown_indices, as PS will initialize the embedding vectors if not exist.